HW2/P3/mandelbrot.pyx

```diff
@@ -3,45 +3,99 @@ cimport numpy as np
+            in_coords_imag[i,j]=in_coords[i,j].imag
+    with nogil, parallel(num_threads=4):
+        #setup thread local buffer
+        out_counts_temp = <float *> malloc(sizeof(float) * 8)
+        for i in prange(in_coords.shape[0],schedule='static',chunksize=1):           
+            for j in range(in_coords.shape[1]/8):
+                c_real = AVX.make_float8(in_coords_real[i,j*8+7],
+                                    in_coords_real[i,j*8+6],
+                                    in_coords_real[i,j*8+5],
+                                    in_coords_real[i,j*8+4],
+                                    in_coords_real[i,j*8+3],
+                                    in_coords_real[i,j*8+2],
+                                    in_coords_real[i,j*8+1],
+                                    in_coords_real[i,j*8+0])
+                c_imag = AVX.make_float8(in_coords_imag[i,j*8+7],
+                                    in_coords_imag[i,j*8+6],
+                                    in_coords_imag[i,j*8+5],
+                                    in_coords_imag[i,j*8+4],
+                                    in_coords_imag[i,j*8+3],
+                                    in_coords_imag[i,j*8+2],
+                                    in_coords_imag[i,j*8+1],
+                                    in_coords_imag[i,j*8+0])
+                z_real = AVX.float_to_float8(0.0)
+                z_imag = AVX.float_to_float8(0.0)
+                avx_out_counts=AVX.float_to_float8(0.0)
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+                    # #do the magnitude squared computation in two steps...
+                    magnitude_squared=AVX.mul(z_real,z_real)
+                    magnitude_squared=AVX.fmadd(z_imag,z_imag,magnitude_squared)
+
+                    #mag^2<4, return 1. 
+                    cont_case=AVX.less_than_equal(magnitude_squared,AVX.float_to_float8(4))
+
+                    #if mag^2<4, cont_case will be -NaN, otherwise 0
+                    #When all iterations for all 8 elements are done, 
+                    #all bits of cont_case will be 0.
+                    if AVX.signs(cont_case) == 0:
                         break
-                    z = z * z + c
-                out_counts[i, j] = iter
 
+                    z_real_new = AVX.add(AVX.sub(AVX.mul(z_real,z_real),AVX.mul(z_imag,z_imag)),c_real)
+                    z_imag = AVX.add(AVX.mul(AVX.mul(z_real,z_imag),AVX.float_to_float8(2)),c_imag)
+                    z_real=z_real_new
+                #add 1 to counts from previous iteration for elements that aren't finished. Do this by masking 1 with cont_case, 
+                    avx_out_counts = AVX.add(avx_out_counts,AVX.bitwise_and(AVX.float_to_float8(1.0),cont_case))
+                AVX.to_mem(avx_out_counts,out_counts_temp)
+                for jj in range(8):
+                    out_counts[i, 8*j+jj] = <int> out_counts_temp[jj]
```

Comment:
I'm not sure why you needed the for loop here after calling AVX.to_mem, wouldn't to_mem put the content in out_counts_temp into avx_out_counts?